pkg/routeagent_driver/handlers/ovn/connection.go

```diff
@@ -180,20 +168,135 @@ func (c *ConnectionHandler) createLibovsdbClient(dbAddress string, tlsConfig *tl
+	certFile, err := clusterfiles.Get(k8sClientset, getOVNCertPath())
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting config for %q", getOVNCertPath())
+	}
+
+	pkFile, err := clusterfiles.Get(k8sClientset, getOVNPrivKeyPath())
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting config for %q", getOVNPrivKeyPath())
+	}
+
+	caFile, err := clusterfiles.Get(k8sClientset, getOVNCaBundlePath())
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting config for %q", getOVNCaBundlePath())
+	}
+
+	tlsConfig, err := getOVNTLSConfig(pkFile, certFile, caFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting OVN TLS config")
+	}
+
+	return tlsConfig, nil
+}
+
+func discoverOvnKubernetesNetwork(ctx context.Context, k8sClientset clientset.Interface, zoneName string) (string, error) {
+	ovnDBPod, err := FindPod(ctx, k8sClientset, "name=ovnkube-db")
+	if err != nil {
+		return "", err
+	}
+
+	var nbdbAddress string
+
+	if ovnDBPod != nil {
+		nbdbAddress, err = discoverOvnDBClusterNetwork(ctx, k8sClientset, ovnDBPod)
+	} else {
+		nbdbAddress, err = discoverOvnNodeClusterNetwork(ctx, k8sClientset, zoneName)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	return nbdbAddress, nil
+}
+
+func discoverOvnDBClusterNetwork(ctx context.Context, k8sClientset clientset.Interface, ovnDBPod *corev1.Pod) (string, error) {
+	_, err := k8sClientset.CoreV1().Services(ovnDBPod.Namespace).Get(ctx, ovnKubeService, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("error finding %q service in %q namespace", ovnKubeService, ovnDBPod.Namespace)
+	}
+
+	dbConnectionProtocol := findProtocol(ovnDBPod)
+
+	return fmt.Sprintf("%s:%s.%s:%d", dbConnectionProtocol, ovnKubeService, ovnDBPod.Namespace, OvnNBDBDefaultPort), nil
+}
+
+func discoverOvnNodeClusterNetwork(ctx context.Context, k8sClientset clientset.Interface, zoneName string) (string, error) {
+	// In OVN IC deployments, the ovn DB will be a part of ovnkube-node
+	ovnPod, err := FindPod(ctx, k8sClientset, "app=ovnkube-node")
+	if err != nil || ovnPod == nil {
+		return "", err
+	}
+
+	endpointList, err := findEndpoint(ctx, k8sClientset, ovnPod.Namespace)
+	if err != nil {
+		return "", errors.Wrapf(err, "error retrieving the endpoints from namespace %q", ovnPod.Namespace)
+	}
+
+	var nbdbAddress string
+
+	if endpointList == nil || len(endpointList.Items) == 0 {
+		nbdbAddress = defaultOVNUnixSocket
+	} else {
+		nbdbAddress = createClusterNetworkWithEndpoints(endpointList.Items, zoneName)
+	}
+
+	return nbdbAddress, nil
+}
+
+func createClusterNetworkWithEndpoints(endPoints []corev1.Endpoints, zoneName string) string {
+	for index := range endPoints {
+		for _, subset := range endPoints[index].Subsets {
+			if strings.Contains(endPoints[index].Name, zoneName) {
+				for _, port := range subset.Ports {
+					if strings.Contains(port.Name, "north") {
+						return fmt.Sprintf("%s:%s:%d",
+							port.Protocol, subset.Addresses[0].IP, OvnNBDBDefaultPort)
```

Comment:
Can this IPaddress even point to IPv6 address? If so, we should look for IPv4 and only use that value.